### PR TITLE
Speed-up Server-Side rendering

### PR DESCRIPTION
**Pull request by forki**
* Uses HashSet instead of Set
* Uses a single StringBuilder for rendering

### CODE REVIEW

**Comment by zaaack** [comment]
@forki  Have you tried it with SampleApp? If the console doesn't have any warning or error I think it would be OK.

**Comment by forki** [comment]
what's the process to test?

**Comment by zaaack** [comment]
If you won't to run the SampleApp, you can start it follow these commands:
```sh
cd ./fable-react/Samples/SSRSample/
./build.sh run # or ./build.cmd run on windows
```
Or you can run the benchmark ⚡️ in SampleApp via these:
```sh
cd ./Samples/SSRSample
./build.sh bench # or ./build.cmd bench on windows
```

These steps are actually already contained in [the tutorial](https://github.com/fable-compiler/fable-react/blob/master/docs/server-side-rendering.md#try-the-sample-app), and don't forget to update the benchmark result if you made big progress : )

**Comment by forki** [comment]
@zaaack 

Old:    12091.898 Requests/sec
New:   27797.081 Requests/sec

can you please confirm?

**Comment by forki** [comment]
@zaaack can you please post our stats with that PR? Then I can update the tutorial accordingly

**Comment by forki** [comment]
29673.591

**Comment by zaaack** [comment]
Here are my new numbers:
node: Requests/sec: 8016.032
New: Requests/sec: 12755.102

dotnet: 2.1.4
node: 8.0.0

**Comment by forki** [comment]
can you post teh full output so that I can replace that part? Also can you describe the improvements that you see in plain english (like "it's 2.5x faster than before and 5x faster than node" )? Do you see any?

**Comment by zaaack** [comment]
I'm worried the difference is related to my MacBook, I'll test it again this evening on my Linux laptop.

**Comment by forki** [comment]
no that's fine. I want the full output  of your macbook and just update the existing benchmarks with whatever it now gives for you

**Comment by zaaack** [comment]
The numbers changed a lot, the new result is only 1.5x faster than node, and the old result is even slower. Don't know why, but I'm afraid the numbers are not accurate on my MacBook now. You can change the tutorial from `my laptop` to `a local laptop`.

**Comment by forki** [comment]
I was asking because the node things was now much faster for me as well.

**Comment by zaaack** [comment]
dotnet ./bin/Release/netcoreapp2.0/dotnet.dll
﻿Thread 1 started
Thread 1 render 10000 times used 1562ms
[Single thread] 1562ms    6402.049req/s
Thread 1 started
Thread 4 started
Thread 6 started
Thread 3 started
Thread 3 render 5000 times used 1564ms
Thread 1 render 5000 times used 1566ms
Thread 6 render 5000 times used 1568ms
Thread 4 render 5000 times used 1575ms
[4 tasks] Total: 1568ms    Memory footprint: 31.902MB   Requests/sec: 12755.102
/usr/local/bin/node ./node.js
﻿Master 29101 is running
[Single process] 1809ms    5527.916req/s
Worker 29105: started
Worker 29103: started
Worker 29102: started
Worker 29104: started
Worker 29103: render 5000 times used 2453ms
Worker 29102: render 5000 times used 2492ms
Worker 29105: render 5000 times used 2510ms
Worker 29104: render 5000 times used 2525ms
[4 workers] Total: 2495ms    Memory footprint: 76.033MB    Requests/sec: 8016.032

**Comment by zaaack** [comment]
Perhaps macbook is not a good device for some benchmark, I have some strange benchmark experience on macbook, with chrome. The numbers are unstable between your multiple tests, but they are almost the same on Linux/windows. So I guess we should test it again in a Linux device, or using the CI.

**Comment by forki** [comment]
can you please send a PR to this PR when you have stable numbers? This is 2.5x times faster than before for me. But I have no real comparison to node.

**Comment by zaaack** [comment]
Yeah, I'll test it on my Linux laptop, it should be more stable and reliable since most people will run dotnet application on Linux or Windows servers.

**Comment by zaaack** [comment]
https://github.com/fable-compiler/fable-react/pull/80 

```sh

dotnet ./bin/Release/netcoreapp2.0/dotnet.dll
Thread 1 started
Thread 1 render 160000 times used 25146ms
[Single thread] 25146ms    6362.841req/s
Thread 1 started
Thread 3 started
Thread 4 started
Thread 5 started
Thread 7 started
Thread 8 started
Thread 9 started
Thread 6 started
Thread 3 render 20000 times used 10370ms
Thread 9 render 20000 times used 10409ms
Thread 6 render 20000 times used 10421ms
Thread 4 render 20000 times used 10434ms
Thread 7 render 20000 times used 10445ms
Thread 1 render 20000 times used 10458ms
Thread 5 render 20000 times used 10483ms
Thread 8 render 20000 times used 10497ms
[8 tasks] Total: 10439ms    Memory footprint: 43.266MB   Requests/sec: 15327.139

/usr/local/bin/node ./node.js
Master 3266 is running
[Single process] 33702ms    4747.493req/s
Worker 3275: started
Worker 3281: started
Worker 3303: started
Worker 3289: started
Worker 3304: started
Worker 3296: started
Worker 3320: started
Worker 3317: started
Worker 3281: render 20000 times used 10247ms
Worker 3289: render 20000 times used 10709ms
Worker 3275: render 20000 times used 11048ms
Worker 3303: render 20000 times used 11119ms
Worker 3296: render 20000 times used 11188ms
Worker 3320: render 20000 times used 11359ms
Worker 3304: render 20000 times used 11370ms
Worker 3317: render 20000 times used 11574ms
[8 workers] Total: 11076.75ms    Memory footprint: 200.066MB    Requests/sec: 14444.670

```
Here are the results on my Linux laptop,  v8's SSR becomes really fast, I didn't update nodejs since last test, not sure what speed up the nodejs' SSR performance. :confused:

**Comment by forki** [comment]
this is ready

File: src/Fable.React/Fable.Helpers.ReactServer.fs [hunk 1]
```diff
@@ -698,14 +704,14 @@ let private renderAttrs (attrs: IProp seq) tag =
       | "textarea", DefaultValue v ->
           childHtml <- Some v
       | _, _ ->
-        html.Append(renderHtmlAttr attr) |> ignore
         html.Append(" ") |> ignore
+        renderHtmlAttr html attr
     | :? SVGAttr as attr ->
-      html.Append(renderSVGAttr attr) |> ignore
-      html.Append(" ") |> ignore
+      html.Append(" ")|> ignore
```
Comment (zaaack) [inline src/Fable.React/Fable.Helpers.ReactServer.fs]: I think single spaces can be char, too.

Comment (forki) [reply src/Fable.React/Fable.Helpers.ReactServer.fs]: done

File: src/Fable.React/Fable.Helpers.ReactServer.fs [hunk 2]
```diff
@@ -9,681 +9,687 @@ open Fable.Helpers.React
...
 
-let private voidTags = set ["area"; "base"; "br"; "col"; "embed"; "hr"; "img"; "input"; "keygen"; "link"; "menuitem"; "meta"; "param"; "source"; "track"; "wbr"]
+let private voidTags = System.Collections.Generic.HashSet<_>(["area"; "base"; "br"; "col"; "embed"; "hr"; "img"; "input"; "keygen"; "link"; "menuitem"; "meta"; "param"; "source"; "track"; "wbr"])
 
 // Adapted from https://github.com/facebook/react/blob/37e4329bc81def4695211d6e3795a654ef4d84f5/packages/react-dom/src/server/escapeTextForBrowser.js#L49
-let escapeHtml (str: string) =
-  let escaped = StringBuilder()
+let escapeHtml (sb:StringBuilder) (str: string) =
   let splits = str.Split('"', '\'', '&', '<', '>')
```
Comment (zaaack) [inline src/Fable.React/Fable.Helpers.ReactServer.fs]: Here is a performance question in my head when I implementing `escapeHtml`, is current version faster or split str to a char array and then compare and append each char.

Comment (forki) [reply src/Fable.React/Fable.Helpers.ReactServer.fs]: I dunno. but you can try

Comment (zaaack) [reply src/Fable.React/Fable.Helpers.ReactServer.fs]: The answer is yes: https://github.com/fable-compiler/fable-react/pull/81

**Comment by forki** [comment]
It's now 2.7 times faster than before and uses less memory.

It's also 4.3 times faster than node (single threaded) and 1.8 times faster than node on 8 workers while only using 13% of the memory.

**Comment by alfonsogarciacaro** [comment]
Released as Fable.React 3.1.0 :+1:
